### PR TITLE
PR 3: Client, tests and bugfixes

### PR DESCRIPTION
**Pull request by xacrimon**
This is the last PR and contains the client implementation, a handful of bugfixes and workarounds for bugs in other libraries and an update for all components to use the latest version of the tokio and tonic libraries. I've added documentation and examples to the readme and some tests which are also documented in the readme.

### CODE REVIEW

**Review by knisbet — APPROVED** [review]

File: client/src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+/// A newtype around a vec of strings to allow structopt to parse it.
+#[derive(Debug, Default)]
+pub struct StringList(pub Vec<String>);
+
+impl FromStr for StringList {
+    type Err = Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Infallible> {
+        Ok(StringList(s.split(',').map(|s| s.to_string()).collect()))
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply client/src/cli.rs:19]: Ah, fair point. I'll admit that most effort went towards things that wasn't the CLI as I'd rather spend my time working on stuff that's more relevant to the focus areas.

**Comment by russjones** [comment]
I'm not seeing `stderr`:

```
$ ./client --domain localhost --endpoint https://localhost:7005 --username foo spawn --program-path objdump --args "invalid-args"
spawned job with id f819bc42-95ab-416a-8532-44eef3326266
$ ./client --domain localhost --endpoint https://localhost:7005 --username foo stream-log --stream-type stderr --past-events --uuid f819bc42-95ab-416a-8532-44eef3326266
stderr log:
terminated with exit code 2
```

Expecting to see something like this:
```
Usage: objdump <option(s)> <file(s)>
 Display information from object <file(s)>.
 At least one of the following switches must be given:
  -a, --archive-headers    Display archive header information
  -f, --file-headers       Display the contents of the overall file header
  -p, --private-headers    Display object format specific file header contents
  -P, --private=OPT,OPT... Display object format specific contents
  -h, --[section-]headers  Display the contents of the section headers
  -x, --all-headers        Display the contents of all headers
  -d, --disassemble        Display assembler contents of executable sections
  -D, --disassemble-all    Display assembler contents of all sections
      --disassemble=<sym>  Display assembler contents from <sym>
  -S, --source             Intermix source code with disassembly
      --source-comment[=<txt>] Prefix lines of source code with <txt>
  -s, --full-contents      Display the full contents of all sections requested
  -g, --debugging          Display debug information in object file
  -e, --debugging-tags     Display debug information using ctags style
  -G, --stabs              Display (in raw form) any STABS info in the file
  -W[lLiaprmfFsoRtUuTgAckK] or
  --dwarf[=rawline,=decodedline,=info,=abbrev,=pubnames,=aranges,=macro,=frames,
          =frames-interp,=str,=loc,=Ranges,=pubtypes,
          =gdb_index,=trace_info,=trace_abbrev,=trace_aranges,
          =addr,=cu_index,=links,=follow-links]
                           Display DWARF info in the file
  --ctf=SECTION            Display CTF info from SECTION
  -t, --syms               Display the contents of the symbol table(s)
  -T, --dynamic-syms       Display the contents of the dynamic symbol table
  -r, --reloc              Display the relocation entries in the file
  -R, --dynamic-reloc      Display the dynamic relocation entries in the file
  @<file>                  Read options from <file>
  -v, --version            Display this program's version number
  -i, --info               List object formats and architectures supported
  -H, --help               Display this information
```

**Comment by xacrimon** [comment]
Ah, I spent some time refactoring and I missed to rename two variables before the final commit. Should be working now. Was reading from the wrong buffer.

File: server/src/server/mod.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+mod api;
+mod auth;
+mod tls;
+
+use anyhow::Result;
+use api::ApiCore;
+use protocol::api_server::ApiServer;
+use tonic::transport::{Server, ServerTlsConfig};
+
+const ADDR: &str = "0.0.0.0:7005";
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply server/src/server/mod.rs]: Good point. I suppose firewalls or similar isn't a good thing to assume when you can avoid it. Adjusted.

File: client/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        domain: &str,
+        identity: Identity,
+        server_ca: Certificate,
+    ) -> Result<Self> {
+        let tls = ClientTlsConfig::new()
+            .domain_name(domain)
+            .ca_certificate(server_ca)
+            .identity(identity);
+
+        let channel = Endpoint::from_shared(endpoint.to_string())?
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply client/src/client.rs]: Neither can I. Didn't think too much of it being a potential footgun. Removed.

**Review by fspmarshall — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Pull request merged**